modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C

```diff
@@ -0,0 +1,182 @@
+  params.addParam<std::vector<Real>>(
+      "lower_bounds", std::vector<Real>(), "Spatially constant lower bound for each parameter.");
+  params.addParam<std::vector<Real>>(
+      "upper_bounds", std::vector<Real>(), "Spatially constant upper bound for each parameter.");
+
+  return params;
+}
+
+ParameterMeshOptimization::ParameterMeshOptimization(const InputParameters & parameters)
+  : OptimizationReporterBase(parameters),
+    _parameter_names(getParam<std::vector<ReporterValueName>>("parameter_names")),
+    _nparams(_parameter_names.size()),
+    _nvalues(_nparams),
+    _parameters(_nparams),
+    _gradients(_nparams),
+    _lower_bounds(getParam<std::vector<Real>>("lower_bounds")),
+    _upper_bounds(getParam<std::vector<Real>>("upper_bounds"))
+{
+  const auto & meshes = getParam<std::vector<FileName>>("parameter_meshes");
+  const auto & families = getParam<MultiMooseEnum>("parameter_families");
+  const auto & orders = getParam<MultiMooseEnum>("parameter_orders");
+  const auto & initial_condition = getParam<std::vector<Real>>("initial_condition");
+  const auto & ntimes = getParam<unsigned int>("num_parameter_times");
+
+  // Size checks
+  if (meshes.size() != _nparams)
+    paramError("parameter_meshes", "There must be a mesh associated with each parameter.");
+  if (families.size() > 1 && families.size() != _nparams)
+    paramError("parameter_families", "There must be a family associated with each parameter.");
+  if (orders.size() > 1 && orders.size() != _nparams)
+    paramError("parameter_orders", "There must be an order associated with each parameter.");
+  if (!initial_condition.empty() && initial_condition.size() != _nparams)
+    paramError("initial_condition",
+               "There must be an initial condition associated with each parameter.");
+  if (!_lower_bounds.empty() && _lower_bounds.size() != _nparams)
+    paramError("lower_bounds", "There must be a lower bound associated with each parameter.");
+  if (!_upper_bounds.empty() && _upper_bounds.size() != _nparams)
+    paramError("upper_bounds", "There must be an upper bound associated with each parameter.");
+
+  _ndof = 0;
+  for (const auto & i : make_range(_nparams))
+  {
+    const std::string family =
+        families.size() > 1 ? std::string(families[i]) : std::string(families[0]);
+    const std::string order = orders.size() > 1 ? std::string(orders[i]) : std::string(orders[0]);
+    const FEType fetype(Utility::string_to_enum<Order>(order),
+                        Utility::string_to_enum<FEFamily>(family));
+
+    ParameterMesh pmesh(fetype, meshes[i]);
+    _nvalues[i] = pmesh.size() * ntimes;
+    _ndof += _nvalues[i];
+
+    _parameters[i] =
+        &declareValueByName<std::vector<Real>>(_parameter_names[i], REPORTER_MODE_REPLICATED);
+    _parameters[i]->assign(_nvalues[i], initial_condition.empty() ? 0.0 : initial_condition[i]);
+    _gradients[i] = &declareValueByName<std::vector<Real>>("grad_" + _parameter_names[i],
```

Comment:
_gradients here is declared by a different name than what is used in OptimizationReporter.  Should they be the same?  If so, I need to make sure I do that in the adjoint data structure pr.  I'll change that PR after this is merged.  Should this naming and creating the reporters be moved to the OPtimizationReporterBase constructor?  I'm wondering for my pr.